Micronaut Data overrides built in enum converter causing failures

In Micronaut Launch was trying to upgrade to Micronaut Data 2.4.x which failed at first with a compilation error since it was using `mappedBy` on `@ManyToOne` which was invalidated here:

https://github.com/micronaut-projects/micronaut-data/commit/55992a645e634456bb36144c953d935b05bb3ebb#diff-f29a7e80baff474ab7661d914685b0d95baa72d04e6ec0f3f942247b5325862fR163-R167

It does appear in JPA that this is invalid, so I removed it but then it failed at runtime with:

```
Caused by: java.lang.IllegalArgumentException: No enum constant io.micronaut.starter.options.Language.kotlin
	at java.base/java.lang.Enum.valueOf(Enum.java:240)
	at io.micronaut.data.runtime.intercept.DataInitializer.lambda$new$1(DataInitializer.java:54)
	at io.micronaut.core.convert.DefaultConversionService.convert(DefaultConversionService.java:135)
	at io.micronaut.core.convert.ConversionService.convert(ConversionService.java:93)
```

This is because `DataInitializer` has a less lenient enum converter that doesn't try to convert the string to upper case and try the conversion again.

In  general this relates to the broader issue that we should be removing `DataInitializer` and having a different `ConversionService` instance just for data to avoid these problems since this would likely break other areas of Micronaut as well like configuration properties and the controller layer.

We will need to fix this problem and do a 2.4.1 before 2.5.0 is out

--- a/data-jdbc/src/test/groovy/io/micronaut/data/jdbc/h2/H2EnumsMappingSpec.groovy
+++ b/data-jdbc/src/test/groovy/io/micronaut/data/jdbc/h2/H2EnumsMappingSpec.groovy
@@ -4,8 +4,10 @@ import io.micronaut.context.ApplicationContext
 import io.micronaut.data.annotation.GeneratedValue
 import io.micronaut.data.annotation.Id
 import io.micronaut.data.annotation.MappedEntity
+import io.micronaut.data.annotation.Query
 import io.micronaut.data.annotation.TypeDef
 import io.micronaut.data.jdbc.annotation.JdbcRepository
+import io.micronaut.data.jdbc.runtime.JdbcOperations
 import io.micronaut.data.model.DataType
 import io.micronaut.data.model.PersistentEntity
 import io.micronaut.data.model.query.builder.sql.Dialect
@@ -13,6 +15,7 @@ import io.micronaut.data.model.query.builder.sql.SqlQueryBuilder
 import io.micronaut.data.repository.CrudRepository
 import io.micronaut.test.extensions.spock.annotation.MicronautTest
 import spock.lang.AutoCleanup
+import spock.lang.PendingFeature
 import spock.lang.Shared
 import spock.lang.Specification
 
@@ -20,6 +23,7 @@ import javax.inject.Inject
 import javax.persistence.Entity
 import javax.persistence.EnumType
 import javax.persistence.Enumerated
+import javax.transaction.Transactional
 
 @MicronautTest
 class H2EnumsMappingSpec extends Specification implements H2TestPropertyProvider {
@@ -34,6 +38,23 @@ class H2EnumsMappingSpec extends Specification implements H2TestPropertyProvider
     @Shared
     @Inject
     JpaEnumEntityRepository jpaEnumEntityRepository = applicationContext.getBean(JpaEnumEntityRepository)
+
+    @PendingFeature(reason = "Investigate why custom inserts don't work")
+    void 'test inserts are broken for custom queries'() {
+        expect:
+        enumEntityRepository.insertValue("b", "b", 1)
+    }
+
+    void 'test read lower case enum'() {
+        when:
+        enumEntityRepository.insertValueExplicit("b", "B")
+        def result = enumEntityRepository.findByAsString(MyEnum.B)
+
+        then:
+        result != null
+        result.asDefault == MyEnum.B
+        result.asString == MyEnum.B
+    }
 
     void 'test enums'() {
         given:
@@ -114,11 +135,30 @@ class H2EnumsMappingSpec extends Specification implements H2TestPropertyProvider
 }
 
 @JdbcRepository(dialect = Dialect.H2)
-interface EnumEntityRepository extends CrudRepository<EnumEntity, Long> {
+abstract class EnumEntityRepository implements CrudRepository<EnumEntity, Long> {
+    private final JdbcOperations jdbcOperations;
 
-    int update(@Id Long id, MyEnum asDefault, MyEnum asString, MyEnum asInt)
+    EnumEntityRepository(JdbcOperations jdbcOperations) {
+        this.jdbcOperations = jdbcOperations
+    }
 
-    Optional<EnumEntity> find(MyEnum asDefault, MyEnum asString, MyEnum asInt)
+    @Transactional
+    void insertValueExplicit(String a, b) {
+        jdbcOperations.prepareStatement("INSERT INTO ENUM_ENTITY(as_default, as_string, as_int) VALUES(?,?,1)", {
+            it.setString(1, a)
+            it.setString(2, b)
+            it.execute()
+        })
+    }
+
+    @Query("INSERT INTO ENUM_ENTITY(as_default, as_string, as_int) VALUES(:asDefault,:asString, :asString)")
+    abstract void insertValue(String asDefault, String asString, int asInt)
+
+    abstract EnumEntity findByAsString(MyEnum value)
+
+    abstract int update(@Id Long id, MyEnum asDefault, MyEnum asString, MyEnum asInt)
+
+    abstract Optional<EnumEntity> find(MyEnum asDefault, MyEnum asString, MyEnum asInt)
 
 }
 

--- a/data-runtime/src/main/java/io/micronaut/data/runtime/intercept/DataInitializer.java
+++ b/data-runtime/src/main/java/io/micronaut/data/runtime/intercept/DataInitializer.java
@@ -51,8 +51,6 @@ class DataInitializer {
             }
             return Optional.of(enumConstants[i]);
         });
-        conversionService.addConverter(String.class, Enum.class, (name, targetType, context) -> Optional.of(Enum.valueOf(targetType, name)));
-
         conversionService.addConverter(Number.class, Character.class, number -> (char) number.intValue());
 
         conversionService.addConverter(byte[].class, UUID.class, UUID::nameUUIDFromBytes);

